### PR TITLE
Store watch revisions and use them on start

### PR DESCRIPTION
**Pull request by yudai**
Not to get all entries on ETCD, store last seen revisions to ETCD and
start watching from them on boot.

### CODE REVIEW

**Comment by yudai** [comment]
will add tests later.

**Comment by yudai** [comment]
cc / @marcin-ptaszynski

File: server/sync_watch.go [hunk 1]
```diff
@@ -51,13 +54,15 @@ func startSyncWatchProcess(server *Server) {
...
+	responseChans := make(map[string]chan *gohan_sync.Event)
 	stopChan := make(chan bool)
 	for _, path := range watch {
+		responseChans[path] = make(chan *gohan_sync.Event)
 		go func(path string) {
+			responseChan := responseChans[path]
 			defer util.LogFatalPanic(log)
 			for server.running {
-				lockKey := lockPath + "/watch"
+				lockKey := lockPath + "/watch" + path
```
Comment (marcin-ptaszynski) [inline server/sync_watch.go]: nit: [path.Join](https://golang.org/pkg/path/#Join)?

File: server/sync_watch.go [hunk 2]
```diff
@@ -67,32 +72,50 @@ func startSyncWatchProcess(server *Server) {
 				defer func() {
 					server.sync.Unlock(lockKey)
 				}()
-				err = server.sync.Watch(path, responseChan, stopChan, gohan_sync.RevisionCurrent)
+
+				fromRevision := int64(gohan_sync.RevisionCurrent)
+				lastSeen, err := server.sync.Fetch(SyncWatchRevisionPrefix + path)
```
Comment (marcin-ptaszynski) [inline server/sync_watch.go]: ditto

File: server/sync_watch.go [hunk 3]
```diff
@@ -67,32 +72,50 @@ func startSyncWatchProcess(server *Server) {
...
-							env := extensions[event]
-							runExtensionOnSync(server, response, env.Clone())
-							return
+	for _, path := range watch {
+		go func(path string) {
+			defer util.LogFatalPanic(log)
+			responseChan := responseChans[path]
+			for server.running {
+				response := <-responseChan
+				err := server.sync.Update(SyncWatchRevisionPrefix+path, strconv.FormatInt(response.Revision, 10))
```
Comment (marcin-ptaszynski) [inline server/sync_watch.go]: ditto

File: server/sync_watch.go (hunk 1 above)
Comment (yudai) [inline server/sync_watch.go]: etcdv3's key names are simple strings and etcd doesn't handle key names as slash separated strings. So simple concatenating would be better I think.

**Comment by yudai** [comment]
Hmm, it's pretty hard to test this in the current test suite, because it doesn't use `RunServer()`. Any good idea to run watch process?

**Comment by nati** [comment]
We may make RunServer easy with https://golang.org/pkg/net/http/httptest/ . but I admit this is beyond scope of this one. so I'm ok for having separate patch for future testing.

**Comment by yudai** [comment]
ok, rebased the commits anyaway.

**Pull request merged**